File: src/Kotlin/FingerTree/src/tools/datastructures/fingertree/BrodalOkasakiHeap.kt
package tools.datastructures.fingertree

import java.util.IdentityHashMap
import kotlin.math.max

/** Statistics returned by the full bootstrapped skew-binomial invariant audit. */
public data class BrodalOkasakiHeapStatistics(
    val count: Int,
    val rootForestLength: Int,
    val maximumRank: Int,
    val maximumDepth: Int,
)

/** One successful minimum deletion, keeping nullable heap elements unambiguous. */
public data class BrodalMinimumView<T>(
    val minimum: T,
    val remainder: BrodalOkasakiHeap<T>,
)

/**
 * An immutable Brodal-Okasaki bootstrapped skew-binomial min-heap.
 *
 * [insert], [meld], and [minimum] use worst-case O(1) comparisons and structural work;
 * [deleteMinimum] is worst-case O(log n). Every update retains prior versions. Custom-comparator
 * heaps may meld only when they retain the exact same [Comparator] object.
 */
public class BrodalOkasakiHeap<T> private constructor(
    private val root: Tree<T>?,
    public val count: Int,
    public val comparator: Comparator<in T>,
) : Iterable<T> {
    public companion object {
        /** Creates an empty natural-order heap. Kotlin's natural comparator is a shared singleton. */
        public fun <T : Comparable<T>> empty(): BrodalOkasakiHeap<T> =
            BrodalOkasakiHeap(root = null, count = 0, comparator = naturalOrder())

        /** Creates an empty heap retaining [comparator] by identity. */
        public fun <T> empty(comparator: Comparator<in T>): BrodalOkasakiHeap<T> =
            BrodalOkasakiHeap(root = null, count = 0, comparator)

        /** O(n) bulk construction by repeated worst-case-O(1) insertion. */
        public fun <T : Comparable<T>> from(items: Iterable<T>): BrodalOkasakiHeap<T> =
            from(items, naturalOrder())

        /** O(n) bulk construction retaining [comparator]. */
        public fun <T> from(
            items: Iterable<T>,
            comparator: Comparator<in T>,
        ): BrodalOkasakiHeap<T> {
            var result = empty(comparator)
            for (item in items) {
                result = result.insert(item)
            }
            return result
        }
    }

    public val isEmpty: Boolean
        get() = root == null

    /** The minimum in O(1), including when T itself is nullable. */
    public val minimum: T
        get() {
            val actualRoot = root ?: throw NoSuchElementException("The heap is empty.")
            return actualRoot.value
        }

    /** Returns a successful value wrapper, or null only when the heap is empty. */
    public fun minimumView(): BrodalMinimumView<T>? {
        val actualRoot = root ?: return null
        return BrodalMinimumView(actualRoot.value, deleteMinimum())
    }

    /** Inserts [item] in worst-case O(1). */
    public fun insert(item: T): BrodalOkasakiHeap<T> {
        val actualRoot = root
            ?: return BrodalOkasakiHeap(Tree(0, item, children = null), 1, comparator)
        val updatedRoot = if (lessOrEqual(item, actualRoot.value)) {
            Tree(0, item, Forest(actualRoot, tail = null))
        } else {
            Tree(0, actualRoot.value, skewInsert(Tree(0, item, children = null), actualRoot.children))
        }
        return BrodalOkasakiHeap(updatedRoot, Math.addExact(count, 1), comparator)
    }

    /**
     * Melds two heaps in worst-case O(1). The exact comparator object is representation policy.
     */
    public fun meld(other: BrodalOkasakiHeap<T>): BrodalOkasakiHeap<T> {
        require(comparator === other.comparator) {
            "Heap melding requires the same comparator object."
        }
        val leftRoot = root ?: return other
        val rightRoot = other.root ?: return this
        val updatedRoot = if (lessOrEqual(leftRoot.value, rightRoot.value)) {
            Tree(0, leftRoot.value, skewInsert(rightRoot, leftRoot.children))
        } else {
            Tree(0, rightRoot.value, skewInsert(leftRoot, rightRoot.children))
        }
        return BrodalOkasakiHeap(
            updatedRoot,
            Math.addExact(count, other.count),
            comparator,
        )
    }

    /** Removes the minimum in worst-case O(log n). */
    public fun deleteMinimum(): BrodalOkasakiHeap<T> {
        val actualRoot = root ?: throw NoSuchElementException("The heap is empty.")
        val children = actualRoot.children ?: return empty(comparator)
        val (minimumTree, remainder) = getMinimum(children)
        val split = splitForest(minimumTree.rank, minimumTree.children)
        var merged = skewMeld(skewMeld(split.trees, remainder), split.embeddedForest)
        val zeros = toArray(split.zeros)
        for (index in zeros.lastIndex downTo 0) {
            merged = skewInsert(zeros[index], merged)
        }
        return BrodalOkasakiHeap(
            Tree(0, minimumTree.value, merged),
            Math.subtractExact(count, 1),
            comparator,
        )
    }

    /** Structural-order iteration is O(n), stack-safe, and performs no comparisons. */
    override fun iterator(): Iterator<T> = object : Iterator<T> {
        private val pending = ArrayList<Tree<T>>().apply { root?.let(::add) }

        override fun hasNext(): Boolean = pending.isNotEmpty()

        override fun next(): T {
            if (pending.isEmpty()) {
                throw NoSuchElementException()
            }
            val tree = pending.removeAt(pending.lastIndex)
            var forest = tree.children
            while (forest != null) {
                pending.add(forest.head)
                forest = forest.tail
            }
            return tree.value
        }
    }

    /**
     * Validates the global root, every fused primitive-child/embedded-forest boundary, skew ranks,
     * heap order, logical count, and depth with an explicit O(n)-space worklist.
     */
    public fun validateStructure(): BrodalOkasakiHeapStatistics {
        val actualRoot = root
        if (actualRoot == null) {
            check(count == 0) { "An empty Brodal-Okasaki heap has a nonzero count." }
            return BrodalOkasakiHeapStatistics(0, 0, 0, 0)
        }
        check(actualRoot.rank == 0) { "The Brodal-Okasaki global root must have rank zero." }
        val rootForestLength = validateSkewForest(actualRoot.children)
        val pending = ArrayList<TreeFrame<T>>()
        pending.add(TreeFrame(actualRoot, depth = 1))
        var logicalCount = 0
        var maximumRank = 0
        var maximumDepth = 0
        while (pending.isNotEmpty()) {
            val frame = pending.removeAt(pending.lastIndex)
            val tree = frame.tree
            check(tree.rank >= 0) { "A Brodal-Okasaki tree has a negative rank." }
            logicalCount = Math.addExact(logicalCount, 1)
            maximumRank = max(maximumRank, tree.rank)
            maximumDepth = max(maximumDepth, frame.depth)
            validateSkewForest(validateFusedChildren(tree))

            var forest = tree.children
            while (forest != null) {
                check(lessOrEqual(tree.value, forest.head.value)) {
                    "A Brodal-Okasaki child outranks its parent."
                }
                pending.add(TreeFrame(forest.head, Math.addExact(frame.depth, 1)))
                forest = forest.tail
            }
        }
        check(logicalCount == count) {
            "Brodal-Okasaki logical count disagrees with its tree graph."
        }
        return BrodalOkasakiHeapStatistics(count, rootForestLength, maximumRank, maximumDepth)
    }

    /** Counts distinct tree objects shared by identity; intended for persistence audits. */
    internal fun sharedTreeCount(other: BrodalOkasakiHeap<T>): Int {
        val oldTrees = IdentityHashMap<Tree<T>, Boolean>()
        collectTrees(root, oldTrees)
        val visited = IdentityHashMap<Tree<T>, Boolean>()
        val pending = ArrayList<Tree<T>>().apply { other.root?.let(::add) }
        var shared = 0
        while (pending.isNotEmpty()) {
            val tree = pending.removeAt(pending.lastIndex)
            if (visited.put(tree, true) != null) {
                continue
            }
            if (oldTrees.containsKey(tree)) {
                shared++
            }
            var forest = tree.children
            while (forest != null) {
                pending.add(forest.head)
                forest = forest.tail
            }
        }
        return shared
    }

    internal fun rootIdentityForTesting(): Any? = root

    private fun collectTrees(tree: Tree<T>?, destination: IdentityHashMap<Tree<T>, Boolean>) {
        val pending = ArrayList<Tree<T>>().apply { tree?.let(::add) }
        while (pending.isNotEmpty()) {
            val current = pending.removeAt(pending.lastIndex)
            if (destination.put(current, true) != null) {
                continue
            }
            var forest = current.children
            while (forest != null) {
                pending.add(forest.head)
                forest = forest.tail
            }
        }
    }

    private fun lessOrEqual(left: T, right: T): Boolean = comparator.compare(left, right) <= 0

    private fun skewInsert(tree: Tree<T>, forest: Forest<T>?): Forest<T> {
        val tail = forest?.tail
        return if (forest != null && tail != null && forest.head.rank == tail.head.rank) {
            Forest(skewLink(tree, forest.head, tail.head), tail.tail)
        } else {
            Forest(tree, forest)
        }
    }

    private fun skewLink(zero: Tree<T>, first: Tree<T>, second: Tree<T>): Tree<T> {
        check(first.rank == second.rank) { "Invalid skew-link ranks." }
        return when {
            lessOrEqual(first.value, zero.value) && lessOrEqual(first.value, second.value) ->
                Tree(first.rank + 1, first.value, Forest(zero, Forest(second, first.children)))
            lessOrEqual(second.value, zero.value) && lessOrEqual(second.value, first.value) ->
                Tree(second.rank + 1, second.value, Forest(zero, Forest(first, second.children)))
            else -> Tree(first.rank + 1, zero.value, Forest(first, Forest(second, zero.children)))
        }
    }

    private fun link(first: Tree<T>, second: Tree<T>): Tree<T> {
        check(first.rank == second.rank) { "Invalid binomial-link ranks." }
        return if (lessOrEqual(first.value, second.value)) {
            Tree(first.rank + 1, first.value, Forest(second, first.children))
        } else {
            Tree(second.rank + 1, second.value, Forest(first, second.children))
        }
    }

    private fun getMinimum(forest: Forest<T>): MinimumTree<T> {
        var minimumForest = forest
        var cursor = forest.tail
        while (cursor != null) {
            if (comparator.compare(cursor.head.value, minimumForest.head.value) < 0) {
                minimumForest = cursor
            }
            cursor = cursor.tail
        }

        val prefix = ArrayList<Tree<T>>()
        cursor = forest
        while (cursor !== minimumForest) {
            prefix.add(cursor!!.head)
            cursor = cursor.tail
        }
        var remainder = minimumForest.tail
        for (index in prefix.lastIndex downTo 0) {
            remainder = Forest(prefix[index], remainder)
        }
        return MinimumTree(minimumForest.head, remainder)
    }

    private fun splitForest(rank: Int, initialForest: Forest<T>?): SplitForest<T> {
        var currentRank = rank
        var zeros: Forest<T>? = null
        var trees: Forest<T>? = null
        var forest = initialForest
        while (true) {
            if (currentRank == 0) {
                return SplitForest(zeros, trees, forest)
            }
            val firstForest = forest
                ?: throw IllegalStateException("Invalid skew-binomial forest invariant.")
            val tail = firstForest.tail
            if (currentRank == 1 && tail == null) {
                return SplitForest(zeros, Forest(firstForest.head, trees), embeddedForest = null)
            }
            tail ?: throw IllegalStateException("Invalid skew-binomial forest invariant.")
            val first = firstForest.head
            val second = tail.head
            val rest = tail.tail
            if (currentRank == 1) {
                return if (second.rank == 0) {
                    SplitForest(Forest(first, zeros), Forest(second, trees), rest)
                } else {
                    SplitForest(zeros, Forest(first, trees), Forest(second, rest))
                }
            }
            if (first.rank == second.rank) {
                return SplitForest(zeros, Forest(first, Forest(second, trees)), rest)
            }
            if (first.rank == 0) {
                zeros = Forest(first, zeros)
                trees = Forest(second, trees)
                forest = rest
            } else {
                trees = Forest(first, trees)
                forest = Forest(second, rest)
            }
            currentRank--
        }
    }

    private fun skewMeld(left: Forest<T>?, right: Forest<T>?): Forest<T>? =
        unionUnique(uniquify(left), uniquify(right))

    private fun uniquify(forest: Forest<T>?): Forest<T>? =
        if (forest == null) null else insertRanked(forest.head, uniquify(forest.tail))

    private fun insertRanked(tree: Tree<T>, forest: Forest<T>?): Forest<T> {
        if (forest == null) {
            return Forest(tree, tail = null)
        }
        check(tree.rank <= forest.head.rank) { "Invalid ranked insertion order." }
        return if (tree.rank < forest.head.rank) {
            Forest(tree, forest)
        } else {
            insertRanked(link(tree, forest.head), forest.tail)
        }
    }

    private fun unionUnique(left: Forest<T>?, right: Forest<T>?): Forest<T>? {
        if (left == null) {
            return right
        }
        if (right == null) {
            return left
        }
        return when {
            left.head.rank < right.head.rank -> Forest(left.head, unionUnique(left.tail, right))
            left.head.rank > right.head.rank -> Forest(right.head, unionUnique(left, right.tail))
            else -> insertRanked(
                link(left.head, right.head),
                unionUnique(left.tail, right.tail),
            )
        }
    }

    private fun validateFusedChildren(tree: Tree<T>): Forest<T>? {
        var rank = tree.rank
        var forest = tree.children
        while (rank > 0) {
            val firstForest = forest
                ?: throw IllegalStateException("A ranked Brodal-Okasaki tree is missing structural children.")
            val first = firstForest.head
            if (rank == 1) {
                check(first.rank == 0) {
                    "A rank-one Brodal-Okasaki tree must begin with a rank-zero child."
                }
                val tail = firstForest.tail ?: return null
                return if (tail.head.rank == 0) tail.tail else tail
            }
            val tail = firstForest.tail
                ?: throw IllegalStateException("A ranked Brodal-Okasaki tree has an incomplete child encoding.")
            val second = tail.head
            if (first.rank == second.rank) {
                check(first.rank == rank - 1) {
                    "A skew-linked Brodal-Okasaki tree has invalid child ranks."
                }
                return tail.tail
            }
            if (first.rank == 0) {
                check(second.rank == rank - 1) {
                    "A skew-linked Brodal-Okasaki tree has an invalid ranked child."
                }
                forest = tail.tail
            } else {
                check(first.rank == rank - 1) {
                    "A linked Brodal-Okasaki tree has an invalid ranked child."
                }
                forest = tail
            }
            rank--
        }
        return forest
    }

    private fun validateSkewForest(initialForest: Forest<T>?): Int {
        var length = 0
        var previousRank = -1
        var duplicate = false
        var forest = initialForest
        while (forest != null) {
            val rank = forest.head.rank
            check(rank >= 0) { "A Brodal-Okasaki forest contains a negative rank." }
            check(rank >= previousRank) {
                "Brodal-Okasaki root-forest ranks are not nondecreasing."
            }
            if (rank == previousRank) {
                check(length == 1 && !duplicate) {
                    "Only the first two skew-forest ranks may be equal."
                }
                duplicate = true
            }
            previousRank = rank
            length++
            forest = forest.tail
        }
        return length
    }

    private fun toArray(forest: Forest<T>?): List<Tree<T>> {
        val result = ArrayList<Tree<T>>()
        var cursor = forest
        while (cursor != null) {
            result.add(cursor.head)
            cursor = cursor.tail
        }
        return result
    }

    private class Tree<T>(
        val rank: Int,
        val value: T,
        val children: Forest<T>?,
    )

    private class Forest<T>(
        val head: Tree<T>,
        val tail: Forest<T>?,
    )

    private data class TreeFrame<T>(val tree: Tree<T>, val depth: Int)

    private data class MinimumTree<T>(val tree: Tree<T>, val remainder: Forest<T>?)

    private data class SplitForest<T>(
        val zeros: Forest<T>?,
        val trees: Forest<T>?,
        val embeddedForest: Forest<T>?,
    )
}

File: src/CSharp/src/Durable7.FingerTree/BrodalOkasakiHeap.cs
// Represents an immutable bootstrapped skew-binomial min-heap.

using System.Collections;
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;

namespace Durable7.FingerTree;

/// <summary>Represents an immutable bootstrapped skew-binomial min-heap.</summary>
/// <typeparam name="T">The element type.</typeparam>
/// <remarks>
/// This is the Brodal–Okasaki purely functional priority queue: insert, minimum, and meld are O(1)
/// worst-case; delete-min is O(log n) worst-case. Every update preserves prior heap versions.
/// </remarks>
[DebuggerDisplay("Count = {Count}, Minimum = {MinimumForDebugger}")]
public sealed class BrodalOkasakiHeap<T> : IEnumerable<T>
{
    private readonly Tree? _root;

    private BrodalOkasakiHeap(Tree? root, int count, IComparer<T> comparer)
    {
        _root = root;
        Count = count;
        Comparer = comparer;
    }

    /// <summary>Gets the shared empty heap using the default comparer.</summary>
    public static BrodalOkasakiHeap<T> Empty { get; } = new(null, 0, Comparer<T>.Default);

    /// <summary>Gets the number of elements.</summary>
    public int Count { get; }

    /// <summary>Gets whether the heap is empty.</summary>
    public bool IsEmpty => _root is null;

    /// <summary>Gets the retained priority comparer.</summary>
    public IComparer<T> Comparer { get; }

    /// <summary>Gets the minimum element.</summary>
    /// <exception cref="InvalidOperationException">The heap is empty.</exception>
    public T Minimum => _root is null
        ? throw new InvalidOperationException("The heap is empty.")
        : _root.Value;

    private object? MinimumForDebugger => _root is null ? null : _root.Value;

    internal object? RootIdentity => _root;

    /// <summary>
    /// Validates the global root, every fused primitive-child/embedded-forest boundary,
    /// skew-forest ranks, heap order, and logical count in O(n) time.
    /// </summary>
    /// <returns>Statistics for the validated bootstrapped skew-binomial representation.</returns>
    /// <remarks>The traversal uses an explicit stack requiring O(n) space in the worst case.</remarks>
    public BrodalOkasakiHeapStatistics ValidateStructure()
    {
        if (_root is null)
        {
            if (Count != 0)
                throw new InvalidOperationException("An empty Brodal-Okasaki heap has a nonzero count.");
            return new BrodalOkasakiHeapStatistics(0, 0, 0, 0);
        }
        if (_root.Rank != 0)
            throw new InvalidOperationException("The Brodal-Okasaki global root must have rank zero.");
        var rootForestLength = ValidateSkewForest(_root.Children);
        var pending = new Stack<TreeFrame>();
        pending.Push(new TreeFrame(_root, 1));
        var logicalCount = 0;
        var maximumRank = 0;
        var maximumDepth = 0;
        while (pending.TryPop(out var frame))
        {
            var tree = frame.Tree;
            if (tree.Rank < 0)
                throw new InvalidOperationException("A Brodal-Okasaki tree has a negative rank.");
            logicalCount = checked(logicalCount + 1);
            maximumRank = Math.Max(maximumRank, tree.Rank);
            maximumDepth = Math.Max(maximumDepth, frame.Depth);
            ValidateSkewForest(ValidateFusedChildren(tree));
            for (var forest = tree.Children; forest is not null; forest = forest.Tail)
            {
                if (!LessOrEqual(tree.Value, forest.Head.Value))
                    throw new InvalidOperationException("A Brodal-Okasaki child outranks its parent.");
                pending.Push(new TreeFrame(forest.Head, checked(frame.Depth + 1)));
            }
        }
        if (logicalCount != Count)
            throw new InvalidOperationException("Brodal-Okasaki logical count disagrees with its tree graph.");
        return new BrodalOkasakiHeapStatistics(Count, rootForestLength, maximumRank, maximumDepth);
    }

    /// <summary>Creates an empty heap with a retained comparer.</summary>
    /// <param name="comparer">The comparer, or <see langword="null"/> for the default.</param>
    /// <returns>An empty heap using the comparer.</returns>
    public static BrodalOkasakiHeap<T> Create(IComparer<T>? comparer = null)
    {
        comparer ??= Comparer<T>.Default;
        return ReferenceEquals(comparer, Comparer<T>.Default) ? Empty : new BrodalOkasakiHeap<T>(null, 0, comparer);
    }

    /// <summary>Creates a heap from a sequence in O(n) time by repeated worst-case-O(1) insertion.</summary>
    /// <param name="items">The items to insert.</param>
    /// <param name="comparer">The comparer, or <see langword="null"/> for the default.</param>
    /// <returns>A heap containing the items.</returns>
    public static BrodalOkasakiHeap<T> CreateRange(IEnumerable<T> items, IComparer<T>? comparer = null)
    {
        ArgumentNullException.ThrowIfNull(items);
        var result = Create(comparer);
        foreach (var item in items)
            result = result.Insert(item);
        return result;
    }

    /// <summary>Tries to get the minimum element in O(1) time.</summary>
    /// <param name="minimum">The minimum on success; otherwise, the default value.</param>
    /// <returns><see langword="true"/> when nonempty.</returns>
    public bool TryGetMinimum([MaybeNullWhen(false)] out T minimum)
    {
        if (_root is null)
        {
            minimum = default;
            return false;
        }
        minimum = _root.Value;
        return true;
    }

    /// <summary>Inserts an element in O(1) worst-case time.</summary>
    /// <param name="item">The item to insert.</param>
    /// <returns>The updated heap.</returns>
    public BrodalOkasakiHeap<T> Insert(T item)
    {
        if (_root is null)
            return new BrodalOkasakiHeap<T>(new Tree(0, item, null), 1, Comparer);
        var root = LessOrEqual(item, _root.Value)
            ? new Tree(0, item, new Forest(_root, null))
            : new Tree(0, _root.Value, SkewInsert(new Tree(0, item, null), _root.Children));
        return new BrodalOkasakiHeap<T>(root, checked(Count + 1), Comparer);
    }

    /// <summary>Melds two heaps in O(1) worst-case time.</summary>
    /// <param name="other">The other heap.</param>
    /// <returns>The melded heap.</returns>
    /// <exception cref="ArgumentException">The heaps do not retain the same comparer object.</exception>
    public BrodalOkasakiHeap<T> Meld(BrodalOkasakiHeap<T> other)
    {
        ArgumentNullException.ThrowIfNull(other);
        if (!ReferenceEquals(Comparer, other.Comparer))
            throw new ArgumentException("Heap melding requires the same comparer object.", nameof(other));
        if (_root is null)
            return other;
        if (other._root is null)
            return this;

        var root = LessOrEqual(_root.Value, other._root.Value)
            ? new Tree(0, _root.Value, SkewInsert(other._root, _root.Children))
            : new Tree(0, other._root.Value, SkewInsert(_root, other._root.Children));
        return new BrodalOkasakiHeap<T>(root, checked(Count + other.Count), Comparer);
    }

    /// <summary>Removes the minimum element in O(log n) worst-case time.</summary>
    /// <returns>The remaining heap.</returns>
    /// <exception cref="InvalidOperationException">The heap is empty.</exception>
    public BrodalOkasakiHeap<T> DeleteMinimum()
    {
        if (_root is null)
            throw new InvalidOperationException("The heap is empty.");
        if (_root.Children is null)
            return Create(Comparer);

        var (minimum, remainder) = GetMinimum(_root.Children);
        var (zeros, trees, forest) = SplitForest(minimum.Rank, null, null, minimum.Children);
        var merged = SkewMeld(SkewMeld(trees, remainder), forest);
        foreach (var zero in ToReverseArray(zeros))
            merged = SkewInsert(zero, merged);
        return new BrodalOkasakiHeap<T>(new Tree(0, minimum.Value, merged), checked(Count - 1), Comparer);
    }

    /// <summary>Tries to remove the minimum element in O(log n) worst-case time.</summary>
    /// <param name="minimum">The removed minimum on success.</param>
    /// <param name="remainder">The remaining heap on success; otherwise, this heap.</param>
    /// <returns><see langword="true"/> when nonempty.</returns>
    public bool TryDeleteMinimum(
        [MaybeNullWhen(false)] out T minimum,
        out BrodalOkasakiHeap<T> remainder)
    {
        if (_root is null)
        {
            minimum = default;
            remainder = this;
            return false;
        }
        minimum = _root.Value;
        remainder = DeleteMinimum();
        return true;
    }

    /// <summary>Returns an O(n)-time enumerator over the heap in unspecified structural order.</summary>
    /// <returns>An enumerator.</returns>
    /// <remarks>The enumerator uses an explicit stack and does not compare elements.</remarks>
    public IEnumerator<T> GetEnumerator() => Enumerate().GetEnumerator();

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

    private IEnumerable<T> Enumerate()
    {
        if (_root is null)
            yield break;
        var stack = new Stack<Tree>();
        stack.Push(_root);
        while (stack.TryPop(out var tree))
        {
            yield return tree.Value;
            for (var forest = tree.Children; forest is not null; forest = forest.Tail)
                stack.Push(forest.Head);
        }
    }

    private bool LessOrEqual(T left, T right) => Comparer.Compare(left, right) <= 0;

    private Forest SkewInsert(Tree tree, Forest? forest)
    {
        if (forest is { Tail: { } tail } && forest.Head.Rank == tail.Head.Rank)
            return new Forest(SkewLink(tree, forest.Head, tail.Head), tail.Tail);
        return new Forest(tree, forest);
    }

    private Tree SkewLink(Tree zero, Tree first, Tree second)
    {
        Debug.Assert(first.Rank == second.Rank);
        if (LessOrEqual(first.Value, zero.Value) && LessOrEqual(first.Value, second.Value))
            return new Tree(first.Rank + 1, first.Value, new Forest(zero, new Forest(second, first.Children)));
        if (LessOrEqual(second.Value, zero.Value) && LessOrEqual(second.Value, first.Value))
            return new Tree(second.Rank + 1, second.Value, new Forest(zero, new Forest(first, second.Children)));
        return new Tree(first.Rank + 1, zero.Value, new Forest(first, new Forest(second, zero.Children)));
    }

    private Tree Link(Tree first, Tree second)
    {
        Debug.Assert(first.Rank == second.Rank);
        return LessOrEqual(first.Value, second.Value)
            ? new Tree(first.Rank + 1, first.Value, new Forest(second, first.Children))
            : new Tree(second.Rank + 1, second.Value, new Forest(first, second.Children));
    }

    private (Tree Minimum, Forest? Remainder) GetMinimum(Forest forest)
    {
        if (forest.Tail is null)
            return (forest.Head, null);
        var (minimum, remainder) = GetMinimum(forest.Tail);
        return LessOrEqual(forest.Head.Value, minimum.Value)
            ? (forest.Head, forest.Tail)
            : (minimum, new Forest(forest.Head, remainder));
    }

    private (Forest? Zeros, Forest? Trees, Forest? Forest) SplitForest(
        int rank,
        Forest? zeros,
        Forest? trees,
        Forest? forest)
    {
        if (rank == 0)
            return (zeros, trees, forest);
        if (forest is null)
            throw new InvalidOperationException("Invalid skew-binomial forest invariant.");
        if (rank == 1 && forest.Tail is null)
            return (zeros, new Forest(forest.Head, trees), null);
        if (forest.Tail is null)
            throw new InvalidOperationException("Invalid skew-binomial forest invariant.");

        var first = forest.Head;
        var second = forest.Tail.Head;
        var rest = forest.Tail.Tail;
        if (rank == 1)
        {
            return second.Rank == 0
                ? (new Forest(first, zeros), new Forest(second, trees), rest)
                : (zeros, new Forest(first, trees), new Forest(second, rest));
        }
        if (first.Rank == second.Rank)
            return (zeros, new Forest(first, new Forest(second, trees)), rest);
        return first.Rank == 0
            ? SplitForest(rank - 1, new Forest(first, zeros), new Forest(second, trees), rest)
            : SplitForest(rank - 1, zeros, new Forest(first, trees), new Forest(second, rest));
    }

    private Forest? SkewMeld(Forest? left, Forest? right) =>
        UnionUnique(Uniquify(left), Uniquify(right));

    private Forest? Uniquify(Forest? forest) =>
        forest is null ? null : InsertRanked(forest.Head, Uniquify(forest.Tail));

    private Forest InsertRanked(Tree tree, Forest? forest)
    {
        if (forest is null)
            return new Forest(tree, null);
        if (tree.Rank < forest.Head.Rank)
            return new Forest(tree, forest);
        return InsertRanked(Link(tree, forest.Head), forest.Tail);
    }

    private Forest? UnionUnique(Forest? left, Forest? right)
    {
        if (left is null)
            return right;
        if (right is null)
            return left;
        if (left.Head.Rank < right.Head.Rank)
            return new Forest(left.Head, UnionUnique(left.Tail, right));
        if (left.Head.Rank > right.Head.Rank)
            return new Forest(right.Head, UnionUnique(left, right.Tail));
        return InsertRanked(Link(left.Head, right.Head), UnionUnique(left.Tail, right.Tail));
    }

    private static Tree[] ToReverseArray(Forest? forest)
    {
        var trees = new List<Tree>();
        for (; forest is not null; forest = forest.Tail)
            trees.Add(forest.Head);
        trees.Reverse();
        return [.. trees];
    }

    private static Forest? ValidateFusedChildren(Tree tree)
    {
        var rank = tree.Rank;
        var forest = tree.Children;
        while (rank > 0)
        {
            if (forest is null)
                throw new InvalidOperationException("A ranked Brodal-Okasaki tree is missing structural children.");

            var first = forest.Head;
            if (rank == 1)
            {
                if (first.Rank != 0)
                    throw new InvalidOperationException("A rank-one Brodal-Okasaki tree must begin with a rank-zero child.");
                if (forest.Tail is null)
                    return null;

                // The fused representation stores primitive children c followed by the embedded
                // heap forest f. When f also starts at rank zero, the paper permits that zero to
                // be decoded as the second primitive child; SplitForest makes the same choice.
                return forest.Tail.Head.Rank == 0 ? forest.Tail.Tail : forest.Tail;
            }
            if (forest.Tail is null)
                throw new InvalidOperationException("A ranked Brodal-Okasaki tree has an incomplete child encoding.");

            var second = forest.Tail.Head;
            if (first.Rank == second.Rank)
            {
                if (first.Rank != rank - 1)
                    throw new InvalidOperationException("A skew-linked Brodal-Okasaki tree has invalid child ranks.");
                return forest.Tail.Tail;
            }

            if (first.Rank == 0)
            {
                if (second.Rank != rank - 1)
                    throw new InvalidOperationException("A skew-linked Brodal-Okasaki tree has an invalid ranked child.");
                forest = forest.Tail.Tail;
            }
            else
            {
                if (first.Rank != rank - 1)
                    throw new InvalidOperationException("A linked Brodal-Okasaki tree has an invalid ranked child.");
                forest = forest.Tail;
            }
            rank--;
        }
        return forest;
    }

    private static int ValidateSkewForest(Forest? forest)
    {
        var length = 0;
        var previousRank = -1;
        var duplicate = false;
        for (; forest is not null; forest = forest.Tail)
        {
            var rank = forest.Head.Rank;
            if (rank < 0)
                throw new InvalidOperationException("A Brodal-Okasaki forest contains a negative rank.");
            if (rank < previousRank)
                throw new InvalidOperationException("Brodal-Okasaki root-forest ranks are not nondecreasing.");
            if (rank == previousRank)
            {
                if (length != 1 || duplicate)
                    throw new InvalidOperationException("Only the first two skew-forest ranks may be equal.");
                duplicate = true;
            }
            previousRank = rank;
            length++;
        }
        return length;
    }

    private readonly record struct TreeFrame(Tree Tree, int Depth);

    private sealed class Tree(int rank, T value, Forest? children)
    {
        internal int Rank { get; } = rank;
        internal T Value { get; } = value;
        internal Forest? Children { get; } = children;
    }

    private sealed class Forest(Tree head, Forest? tail)
    {
        internal Tree Head { get; } = head;
        internal Forest? Tail { get; } = tail;
    }
}

/// <summary>Describes a validated bootstrapped skew-binomial heap representation.</summary>
/// <param name="Count">The logical element count.</param>
/// <param name="RootForestLength">The number of trees directly below the global minimum root.</param>
/// <param name="MaximumRank">The largest skew-binomial rank.</param>
/// <param name="MaximumDepth">The deepest global-root-to-tree path.</param>
public readonly record struct BrodalOkasakiHeapStatistics(
    int Count,
    int RootForestLength,
    int MaximumRank,
    int MaximumDepth);
